src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2024 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// the corresponding header
+#include "dbgroup/atomic/mwcas/lock_free/mwcas_descriptor.hpp"
+
+// C++ standard libraries
+#include <atomic>
+#include <cstddef>
+
+// external libraries
+#include "dbgroup/lock/common.hpp"
+
+// local sources
+#include "dbgroup/atomic/mwcas/utility.hpp"
+
+namespace dbgroup::atomic::mwcas::lock_free
+{
+/*##############################################################################
+ * Static utilities
+ *############################################################################*/
+
+auto
+MwCASDescriptor::GetDescriptor()  //
+    -> MwCASDescriptor *
+{
+  /*auto *page = _gc->GetPageIfPossible<MwCASDescriptor>();
+  auto *desc = (page == nullptr) ? new MwCASDescriptor{} : static_cast<MwCASDescriptor *>(page);
+  desc->target_cnt_ = 0;*/
+  // GCを無視した一時的な実装
+  MwCASDescriptor *desc{};
+  return desc;
+}
+
+/*##############################################################################
+ * Utilities
+ *############################################################################*/
+auto
+MwCASDescriptor::MwCAS()  //
+    -> bool
+{
+  // set a memory fence
+  stat_.store(kUndecided, kRelease);
+  const auto succeeded = MwCASInternal();
+  return succeeded;
+}
+
+auto
+MwCASDescriptor::MwCASInternal(const size_t begin_pos)  //
+    -> bool
+{
+  const auto desc_addr = std::bit_cast<uint64_t>(this) | kMwCASFlag;
+  auto cur_stat = stat_.load(kSeqCst);
+  if (cur_stat == kUndecided) {
+    auto stat = kSucceeded;
+    for (size_t i = begin_pos; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      auto *addr = target.addr;
+      auto word = addr->load(kSeqCst);
+      while (((word & kMwCASFlag) > 0) && word != desc_addr) {
+        const auto another_desc_addr = word;
+        std::this_thread::sleep_for(kBackOffTime);
+        word = addr->load(kSeqCst);
+        if (word != another_desc_addr) continue;  // other threads modified this field
+        // a long CPU stall has been detected, so perform another MwCAS
```

Comment:
空行を一切入れずにプログラムを書くと，処理のまとまりがわからなくなるので適宜入れましょう．基本的に，suggestoin入れてるときの空行などはたいてい意図を持って入れてます．

```suggestion

        // a long CPU stall has been detected, so perform another MwCAS
```